MobileWallet/TariLib/Wrappers/Wallet.swift

```diff
@@ -60,41 +60,103 @@ class Wallet {
-        return wallet_get_pending_incoming_balance(ptr)
+        var i: Int32 = -1
+        let err = UnsafeMutablePointer<Int32>(&i)
+        let result = wallet_get_pending_incoming_balance(ptr, err)
+        return result
     }
 
     var pendingOutgoingBalance: UInt64 {
-        return wallet_get_pending_outgoing_balance(ptr)
+        var i: Int32 = -1
+        let err = UnsafeMutablePointer<Int32>(&i)
+        let result = wallet_get_pending_outgoing_balance(ptr, err)
+        return result
     }
 
     var publicKey: PublicKey {
-        return PublicKey(pointer: wallet_get_public_key(ptr))
+        var i: Int32 = -1
+        let err = UnsafeMutablePointer<Int32>(&i)
+        let result = PublicKey(pointer: wallet_get_public_key(ptr, err))
+        return result
     }
 
     init(commsConfig: CommsConfig, loggingFilePath: String) {
         let loggingFilePathPointer = UnsafeMutablePointer<Int8>(mutating: (loggingFilePath as NSString).utf8String)!
 
-        ptr = wallet_create(commsConfig.pointer, loggingFilePathPointer)
+        let callback_received_transaction_fn: (@convention(c) (OpaquePointer?) -> Void)? = {
+            valuePointer in
+            let pendingInbound = PendingInboundTransaction.init(pendingInboundTransactionPointer: valuePointer!)
+            print(pendingInbound)
+            }
+
+        let callback_received_transaction_reply_fn: (@convention(c) (OpaquePointer?) -> Void)? = {
+            valuePointer in
+            let completed = CompletedTransaction.init(completedTransactionPointer: valuePointer!)
+            print(completed)
+            }
+
+        let callback_received_finalized_transaction_fn: (@convention(c) (OpaquePointer?) -> Void)? = {
+            valuePointer in
+            let completed = CompletedTransaction.init(completedTransactionPointer: valuePointer!)
+            print(completed)
+            }
+
+        let callback_transaction_broadcast_fn: (@convention(c) (OpaquePointer?) -> Void)? = {
+            valuePointer in
+            let completed = CompletedTransaction.init(completedTransactionPointer: valuePointer!)
+            print(completed)
+            }
+
+        let callback_transaction_mined_fn: (@convention(c) (OpaquePointer?) -> Void)? = {
+            valuePointer in
+            let completed = CompletedTransaction.init(completedTransactionPointer: valuePointer!)
+            print(completed)
+            }
+
+        let callback_discovery_process_complete_fn: (@convention(c) (UInt64, Bool) -> Void)? = {
+            txID, success in
+            print(txID, success)
+            }
+
+        var i: Int32 = -1
+        let err = UnsafeMutablePointer<Int32>(&i)
+
+        ptr = wallet_create(commsConfig.pointer, loggingFilePathPointer, callback_received_transaction_fn, callback_received_transaction_reply_fn, callback_received_finalized_transaction_fn, callback_transaction_broadcast_fn, callback_transaction_mined_fn, callback_discovery_process_complete_fn, err)
```

Comment:
Do callback functions have to be registered before `wallet_create` now? We can't add one later when the app needs to subscribe to some changes?

Reply:
Yes they do all need to be registered on creation.

Reply:
Ok sort of changes the plan I had for internally subscribing to updates but sure I can find a way around 👍